x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -254,6 +236,50 @@ public void setEnabled(String uid, boolean enabled, RefreshPolicy refreshPolicy,
         doUpdate(buildUpdateRequest(uid, builder, refreshPolicy, -1, -1), listener.map(updateResponse -> AcknowledgedResponse.TRUE));
     }
 
+    // package private for testing
+    SearchRequest buildSearchRequest(SuggestProfilesRequest request) {
+        final BoolQueryBuilder query = QueryBuilders.boolQuery().filter(QueryBuilders.termQuery("user_profile.enabled", true));
+        if (Strings.hasText(request.getName())) {
+            query.must(
+                QueryBuilders.multiMatchQuery(
+                    request.getName(),
+                    "user_profile.user.username",
+                    "user_profile.user.username._2gram",
+                    "user_profile.user.username._3gram",
+                    "user_profile.user.full_name",
+                    "user_profile.user.full_name._2gram",
+                    "user_profile.user.full_name._3gram",
+                    "user_profile.user.email"
+                ).type(MultiMatchQueryBuilder.Type.BOOL_PREFIX).fuzziness(Fuzziness.AUTO)
+            );
+        }
+        final SuggestProfilesRequest.Hint hint = request.getHint();
+        if (hint != null) {
+            if (hint.getUids() != null) {
+                assert false == hint.getUids().isEmpty() : "uids hint cannot be empty";
+                query.should(QueryBuilders.termsQuery("user_profile.uid", hint.getUids()));
+            }
+            if (hint.getLabels() != null) {
+                assert hint.getLabels().size() == 1 : "labels hint support exactly one key";
+                final String labelKey = hint.getLabels().keySet().iterator().next();
```

Comment:
Sanity check: it's the intention here that if we somehow end up with no labels, we throw `NoSuchElelement` here, correct? 

Would consider encapsulating this into a `getSingleLabel()` method on the Hint class, together with the assertion, and possibly throw an `IllegalStateException` with a "labels hint support exactly one key" error message there

Reply:
`labels` itself can be `null` but if it is not `null`, it must have exactly one key/value pair. 

Having `getSingleLabel()` is a good idea. I added it and it also allowed me to remove the more general `getLabels` method.